editor/editor_themes.cpp

```diff
@@ -603,29 +603,47 @@ Ref<Theme> create_editor_theme(const Ref<Theme> p_theme) {
 	Ref<StyleBoxFlat> style_content_panel_vp = style_content_panel->duplicate();
-	style_content_panel_vp->set_default_margin(MARGIN_LEFT, border_width);
+	style_content_panel_vp->set_default_margin(MARGIN_LEFT, border_width * 2);
 	style_content_panel_vp->set_default_margin(MARGIN_TOP, default_margin_size);
-	style_content_panel_vp->set_default_margin(MARGIN_LEFT, border_width);
-	style_content_panel_vp->set_default_margin(MARGIN_BOTTOM, border_width);
+	style_content_panel_vp->set_default_margin(MARGIN_RIGHT, border_width * 2);
+	style_content_panel_vp->set_default_margin(MARGIN_BOTTOM, border_width * 2);
 	theme->set_stylebox("panel", "TabContainer", style_content_panel);
 	theme->set_stylebox("Content", "EditorStyles", style_content_panel_vp);
 
 	// Separators (no separators)
 	theme->set_stylebox("separator", "HSeparator", make_line_stylebox(separator_color, border_width));
 	theme->set_stylebox("separator", "VSeparator", make_line_stylebox(separator_color, border_width, 0, true));
 
-	// Debugger
-	Ref<StyleBoxFlat> style_panel_debugger = style_default->duplicate();
-	theme->set_stylebox("DebuggerPanel", "EditorStyles", style_panel_debugger);
+	// HACK Debuger panel
+	Ref<StyleBoxFlat> style_panel_debugger = style_content_panel->duplicate();
+	const int v_offset = theme->get_font("font", "Tabs")->get_height() + style_tab_selected->get_minimum_size().height + default_margin_size * EDSCALE;
+	style_panel_debugger->set_expand_margin_size(MARGIN_TOP, -v_offset);
+	theme->set_stylebox("debugger_panel", "EditorStyles", style_panel_debugger);
 
+	// Debugger
+	Ref<StyleBoxFlat> style_debugger_contents = style_content_panel->duplicate();
+	style_debugger_contents->set_default_margin(MARGIN_LEFT, 0);
+	style_debugger_contents->set_default_margin(MARGIN_BOTTOM, 0);
+	style_debugger_contents->set_default_margin(MARGIN_RIGHT, 0);
+	style_debugger_contents->set_border_width_all(0);
+	style_debugger_contents->set_expand_margin_size(MARGIN_TOP, -v_offset);
+	theme->set_stylebox("DebuggerPanel", "EditorStyles", style_debugger_contents);
 	Ref<StyleBoxFlat> style_tab_fg_debugger = style_tab_selected->duplicate();
-	style_tab_fg_debugger->set_border_width_all(0);
-
+	style_tab_fg_debugger->set_expand_margin_size(MARGIN_LEFT, default_margin_size * EDSCALE + border_width);
+	style_tab_fg_debugger->set_default_margin(MARGIN_LEFT, tab_default_margin_side - default_margin_size * EDSCALE);
 	theme->set_stylebox("DebuggerTabFG", "EditorStyles", style_tab_fg_debugger);
-	theme->set_stylebox("DebuggerTabBG", "EditorStyles", style_tab_unselected);
+	Ref<StyleBoxFlat> style_tab_bg_debugger = style_tab_unselected->duplicate();
+	style_tab_bg_debugger->set_expand_margin_size(MARGIN_LEFT, default_margin_size * EDSCALE + border_width);
+	style_tab_bg_debugger->set_default_margin(MARGIN_LEFT, tab_default_margin_side - default_margin_size * EDSCALE);
+	theme->set_stylebox("DebuggerTabBG", "EditorStyles", style_tab_bg_debugger);
```

Comment:
@djrm I'm not sure if that is caused by this commit.
but I have weird results in the bottom panel.


<img width="506" alt="screen shot 2017-09-18 at 18 29 59" src="https://user-images.githubusercontent.com/16718859/30553409-6be309bc-9ca1-11e7-8acb-0617a0617ee4.png">
<img width="566" alt="screen shot 2017-09-18 at 18 30 26" src="https://user-images.githubusercontent.com/16718859/30553452-8b35cd68-9ca1-11e7-9b8c-06ec4e1d0a70.png">

<img width="738" alt="screen shot 2017-09-18 at 18 29 51" src="https://user-images.githubusercontent.com/16718859/30553365-50956c0e-9ca1-11e7-869e-0d37b08d6160.png">

What are the exact motivations for that PR? It should fix the bottom panel in a way without touching the debug panel. anything else?

I also was thinking about that, but I think it should not be the themes job, to adapt false offsets. IMO this makes the theme complexes to understand.
Just thoughts...

Reply:
@toger5, interesting, i had bad results before my changes, so i fixed them, now i discover that it causes problems for you.

Reply:
with reverting it works again.
so this is caused by this commit.
maybe I should open an issue?
I'm pretty sure your implementation works well. is it maybe just a hiDpi issue?

Reply:
@toger5 let me check with hidpi, and if there is something i will fix it.

Reply:
@djrm sounds great!!
and another question: do we need another EditorStyle
`theme->set_stylebox("DebuggerPanel", "EditorStyles", style_debugger_contents);`

As I said, I almost think that the custom control position/size is a easier to understand/maintain implementation.
As far as I can tell you move the margin + the position + the expand_margins of the tabs so that they are just a little more to the left and align with the bottom_panel.
This is really smart. I was thinking about a way to do it without touching frames but Couldn't come up with one...
Just moving it makes more sense though. since that makes the theme more mentionable and it still works with any theme settings as expected.